specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,126 @@
 
 ## Overview
 
-// TODO: this is a stub.
+Hooks are a mechanism whereby application developers can add arbitrary behavior
+to flag evaluation. They operate similarly to middleware in many web frameworks.
+
+### Definitions
+
+**Stage**: An explicit portion of the flag evaluation lifecycle. e.g. `before` being "before the evaluation is run.
+**Invocation**: A single call to evaluate a flag. `client.getBooleanValue(..)` is an invocation.
+**API**: The global API singleton.
+
+### Hook context
+
+Hook context exists to provide hooks with information about the invocation.
+
+> Hook context **MUST** provide: the flag key, evaluation context, default value and a list of  executed hooks by stage.
+
+> Condition: You **MUST** provide `flag type` if the language type system differentiates between strings, numbers, booleans, and structures.
+
+> Hook context **SHOULD** provide: provider, client
+
+> flag key, flag type, default value properties **MUST** be immutable.
+
+> The evaluation context **MUST** be mutable only within the `before` hook.
+
+### State
+
+> State **MUST** be a map of objects.
+
+> Condition: State **MUST** be immutable, if the implementation language supports a mechanism for marking data as immutable.
+
+### Hook creation and parameters
+
+> Hooks **MUST** specify at least one stage.
+
+> The `before` stage *MUST* run before flag evaluation occurs. It accepts a `hook context` (required) and `state` (optional) as parameters and returns either a `HookContext` or nothing.
+
+```
+HookContext|void before(HookContext, State)
+```
+
+> The `after` stage **MUST** run after flag evaluation occurs. It accepts a `hook context` (required), `flag evaluation details` (required) and `state` (optional). It has no return value.
+
+> The `error` hook runs when errors are encountered in the `before` or `after` state or when flag evaluation errors. It accepts `hook context` (required), `exception` for what went wrong (required), and `state` (optional). It has no return value.
+
+> The `finally` hook **MUST** run after the `before`, `after`, and `error` stages. It accepts a `hook context` (required) and `state` (optional). There is no return value.
+
+> Condition: If `finally` is a reserved word in the language, `afterAll` should be used.
+
+### Hook registration & ordering
+
+> The API, Client and invocation **MUST** have a method for registering hooks which accepts `flag evaluation options`
+
+```js
+OpenFeature.addHooks(new Hook1());
+
+//...
+
+Client client = OpenFeature.getClient();
+client.addHooks(new Hook2());
+
+//...
+
+client.getValue('my-flag', 'defaultValue', new Hook3());
+```
+
+> Hooks **MUST** be evaluated in the following order: API before, client before, invocation before, invocation after, client after, API after, invocation finally, client finally, API finally. If errors are present, the order is: invocation error, client error, API error.
```

Comment:
for the errors scenario, I think it would be clearly to specify the full sequence, which I _think_ is:

- API before
- client before
- invocation before
- invocation error
- client error
- API error
- invocation finally
- client finally
- API finally